File: python/analysis.py
# Import required modules
import argparse
import face_recognition
import time
import cv2 as cv
import math
from utils.path import list_dirs, list_images
from utils.argument import str2bool
from utils.time import transTime
from utils.face import faceEncoding

############ Add argument parser for command line arguments ############
parser = argparse.ArgumentParser(description="Analysis face data")
parser.add_argument(
    "--dataset",
    type=str,
    default="../server/data/",
    help="path to input directory of faces + images",
)
parser.add_argument(
    "--errors",
    type=str,
    default="./log/error.txt",
    help="path to output error face lists.",
)
parser.add_argument(
    "--log", type=str, default="./log/log.txt", help="path to output log file."
)
parser.add_argument(
    "--detection",
    type=str2bool,
    nargs="?",
    const=True,
    default=False,
    help="Toggle of perform face detection first.",
)
parser.add_argument(
    "--resize",
    type=int,
    nargs="?",
    const=True,
    default=640,
    help="Perform image resizing before learning.",
)
parser.add_argument(
    "--threshold",
    type=float,
    default=0.6,
    help="distance threshold for face recognition.",
)
args = parser.parse_args()


def main():
    # get start time
    start_time = time.time()

    # initialize our lists of extracted facial embeddings and
    # corresponding people names
    print("[INFO] quantifying faces...")
    knownNames = list_dirs(args.dataset)
    print("[INFO] total {} users".format(len(knownNames)))

    # initialize log lists
    log = []

    # initialize error lists
    error = []

    # initial variables
    refFace = []

    for name in knownNames:
        imagePaths = list(list_images(args.dataset + name + "/"))
        info = "[INFO] Check face images of {}".format(name)
        print(info)
        if args.log:
            log.append(info)

        flagRef = True
        for (i, imagePath) in enumerate(imagePaths):
            print("[INFO] processing image {}/{}".format(i + 1, len(imagePaths)))
            if flagRef is True:
                refFace = faceEncoding(args, imagePath)
                if len(refFace) > 0:
                    flagRef = False
                else:
                    print("[ERROR] no face found in image {}".format(imagePath))
                    error.append(imagePath)
            else:
                curFace = faceEncoding(args, imagePath)
                if len(curFace) > 0:
                    face_distances = face_recognition.face_distance(refFace, curFace[0])
                    print(face_distances[0])
                    if args.log:
                        log.append(face_distances[0])
                    """ ADD CONTENT
                    if face_distances[0] > args.threshold:
                    """
                else:
                    print("[ERROR] no face found in image {}".format(imagePath))
                    error.append(imagePath)

    # display and save error list
    if len(error) > 0:
        print("[INFO] total {} error images".format(len(error)))
        with open(args.errors, "w", encoding="utf-8") as f:
            for item in error:
                f.write("%s\n" % item)
            f.close()
            print("[INFO] error lists {} saved".format(args.errors))

    # display and save log
    print("[INFO] total {} face images processed".format(len(log)))
    with open(args.log, "w", encoding="utf-8") as f:
        for item in log:
            f.write("%s\n" % item)
        f.close()
        print("[INFO] log {} saved".format(args.log))

    # calculate processing time
    tick = (time.time() - start_time) * 1000
    transTime(tick, "[INFO] Total process time: ")


if __name__ == "__main__":
    main()

File: node/analysis.ts
import * as faceapi from 'face-api.js'
import { canvas, faceDetectionNet, modelLink, getAll } from './utils'
import * as fs from 'fs'
import * as path from 'path'

async function run() {
  // load face recognition model
  await faceDetectionNet.loadFromDisk(modelLink)
  await faceapi.nets.faceLandmark68Net.loadFromDisk(modelLink)
  await faceapi.nets.faceRecognitionNet.loadFromDisk(modelLink)

  // get start time
  const tstart = process.hrtime()

  // grab the paths to the input images in our dataset
  console.info('[INFO] quantifying faces...')
  const userList = getAll()

  // initialize log lists
  const baseDir = path.resolve(__dirname, './output')
  if (!fs.existsSync(baseDir)) {
    fs.mkdirSync(baseDir)
  }
  const fileName = 'analysis_log_' + new Date().getTime().toString() + '.txt'
  const log = fs.createWriteStream(path.resolve(baseDir, fileName))

  await Promise.all(
    userList.map(async (user: { name: string; files: string[] }) => {
      let refDescriptor = null
      await Promise.all(
        user.files.map(async (file: string, index: number) => {
          // extract the person name from the image path
          const img = await canvas.loadImage('..' + file)
          if (index === 0) {
            // find the face and face encodings in the current image and save as reference
            refDescriptor = (await faceapi.computeFaceDescriptor(
              img
            )) as Float32Array
            console.info('[INFO] processing user ' + user.name)
            log.write('[INFO] processing user ' + user.name + '\n')
          } else {
            // find the face and face encodings in the current image
            const descriptor = (await faceapi.computeFaceDescriptor(
              img
            )) as Float32Array

            // compute the euclidean distance of two images
            const dist = faceapi.euclideanDistance(descriptor, refDescriptor)
            log.write(dist + '\n')
            console.info('[INFO] face distance is ' + dist.toString())
          }
        })
      )
    })
  )

  // save and close log
  log.end()

  const tend = process.hrtime(tstart)
  console.info('[INFO] total process time: %dms', tend[1] / 1000000)
}

run()
